Synonyms Overview Documentation (#98202)
elastic/elasticsearch: docs/reference/analysis/synonyms.asciidoc

[[analysis-synonyms]]
== Synonyms

> Synonyms are words or phrases that have the same or similar meaning.

Reviewer: I think it's OK, but I don't think it's a game changer either way!

They are an important aspect of search, as they can improve the search experience and the relevancy of search results.

Synonyms will allow you to:

* Improve search relevance: Synonyms enable users to find relevant documents even if they use different terms to express the same concept.
* Increase search scope: Using synonyms ensures a wider range of possible terms. Search results are not limited to the specific terms included in your documents, and variations of relevant terms are included in search results.
* Domain-specific vocabulary: Certain domains use specific terms that users are not familiar with. Using synonyms allow search to become more user-friendly, as users can search terms they are familiar with and search results will not be limited to the specific domain vocabulary.

> * Handling common misspellings and typos: Synonyms can be used to make sure that frequent or specific misspellings are correctly handled.

Reviewer: I would say that @kderusso has suggested some nice rephrasings here, that are concise in terms of the language used, rather than removing too much of the original information. I'd feel free to to pick and choose which suggestions to apply, finding a happy medium.

> * Handling common misspellings and typos: Synonyms can be used to make sure that frequent or specific misspellings are correctly handled.

Author: 👍 Thanks for confirming! Adding suggestions

Synonyms are grouped together using *synonyms sets*.
You can have as many synonyms sets as you need.

In order to use synonyms sets in {es}, you need to:

* <<synonyms-define-synonyms>>
* <<synonyms-store-synonyms>>
* <<synonyms-synonym-analyzers>>

[discrete]
[[synonyms-define-synonyms]]
=== Define your synonyms sets

Synonyms in a synonyms set are defined using *synonym rules*.
Each synonym rule contains words that are synonyms.

You can use two formats to define synonym rules: Solr and WordNet.

[discrete]
[[synonyms-define-synonyms-solr]]

> ==== Solr format

Reviewer 2: Instead of describing and repeating the same information, let's link the[ original source](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html#_solr_synonyms):

[[analysis-synonyms]]

> == Synonyms

Reviewer 2: Let's call this  "Search with synonyms"

Synonyms are words or phrases that have the same or similar meaning.

> They are an important aspect of search, as they can improve the search experience and the relevancy of search results.

Reviewer 2: "improve relevancy" = > increase the scope

Usually synonyms reduce relevancy, but increase the recall (scope) of search results.

Author: Thanks Mayya! 💯

Synonyms allow you to:

* Improve search relevance: Synonyms enable users to find relevant documents even if they use different terms to express the same concept.

> * Increase search scope: Using synonyms ensures a wider range of possible terms. Search results are not limited to the specific terms included in your documents, and variations of relevant terms are included in search results.

Reviewer 2: This seems to me just a paraphrase of the previous item, we should combine them into the one.

Synonyms usually improve recall or scope which is one of the component of search relevance.

Author: Sounds good 👍

* Domain-specific vocabulary: Certain domains use specific terms that users are not familiar with. Using synonyms allow search to become more user-friendly, as users can search terms they are familiar with and search results will not be limited to the specific domain vocabulary.
* Handling common misspellings and typos: Synonyms can be used to make sure that frequent or specific misspellings are correctly handled.

> Synonyms are grouped together using *synonyms sets*.

Reviewer 2: Is this supposed to be a general document about "synonyms"?  Why we are talking here specifically about "synonyms set"?

You can have as many synonyms sets as you need.

In order to use synonyms sets in {es}, you need to:

* <<synonyms-define-synonyms>>
* <<synonyms-store-synonyms>>
* <<synonyms-synonym-analyzers>>

[discrete]
[[synonyms-define-synonyms]]
=== Define synonyms sets

Synonyms in a synonyms set are defined using *synonym rules*.
Each synonym rule contains words that are synonyms.

You can use two formats to define synonym rules: Solr and WordNet.

[discrete]
[[synonyms-define-synonyms-solr]]
==== Solr format

This format uses two different definitions:

* Equivalent synonyms: Define groups of words that are equivalent. Words are separated by commas. Example:
[source,synonyms]
----
ipod, i-pod, i pod
computer, pc, laptop
----
* Explicit mappings: Matches a group of words to other words. Words on the left hand side of the rule definition are expanded into all the possibilities described on the right hand side. Example:

[source,synonyms]
----
personal computer => pc
sea biscuit, sea biscit => seabiscuit
----

[discrete]
[[synonyms-define-synonyms-wordnet]]
==== WordNet format

> https://wordnet.princeton.edu/[WordNet] defines synonyms sets spanning multiple lines. Each line contains the following information:

Reviewer 2: I think better to not provide any links, but in case you want to provide we need to provide a link to the wordnet format: https://wordnet.princeton.edu/documentation/prologdb5wn

Author: I haven't found a direct link to the format used here. Checking the link you provided, I don't see that it is a direct description of the format either 😢 

I'm happy to remove the link, but I think it provides an initial link where users can investigate.

* Synonyms set numeric identifier
* Ordinal of the synonym in the synonyms set
* Synonym word
* Word type identifier: Noun (n), verb (v), adjective (a) or adverb (b).
* Depth of the word in the synonym net

The following example defines a synonym set for the words "come", "advance" and "approach":

[source,synonyms]
----
s(100000002,1,'come',v,1,0).
s(100000002,2,'advance',v,1,0).
s(100000002,3,'approach',v,1,0).""";
----

[discrete]
[[synonyms-store-synonyms]]
=== Store your synonyms set

Your synonyms sets need to be stored in {es} so your analyzers can refer to them.

> There are three ways to include your synonyms sets:

Reviewer 2: "to include your synonyms sets"  => "to store your synonyms sets"

[discrete]
[[synonyms-store-synonyms-api]]
==== Synonyms API

You can use the <<synonyms-apis,synonyms APIs>> to manage synonyms sets.
This is the most flexible approach, as it allows to dynamically define and modify synonyms sets.

Changes in your synonyms sets will automatically reload the associated analyzers.

[discrete]
[[synonyms-store-synonyms-file]]
==== Synonyms File

You can store your synonyms set in a file.

A synonyms set file needs to be uploaded to all your cluster nodes, and be located in the configuration directory for your {es} distribution.
If you're using {ess}, you can upload synonyms files using {cloud}/ec-custom-bundles.html[custom bundles].

> 

Reviewer 2: I think it is better to remove this line about ESS. Soon in Serverless there will no option to use synonym files.

Author: ESS won't go away with serverless. I think it's good to keep this line as it can be cumbersome for users to find how to deal with synonyms in ESS.

An example synonyms file:

[source,synonyms]
--------------------------------------------------
include::{es-test-dir}/cluster/config/analysis/synonym.txt[]
--------------------------------------------------

To update an existing synonyms set, upload new files to your cluster.
Synonyms set files must be kept in sync on every cluster node.

When a synonyms set is updated, search analyzers that use it need to be refreshed using the <<indices-reload-analyzers,reload search analyzers API>>

This manual syncing and reloading makes this approach less flexible than using the <<synonyms-store-synonyms-api,synonyms API>>.

[discrete]
[[synonyms-store-synonyms-inline]]
==== Inline

> For testing out your synonyms, you can directly inline them in your token filter definition.

Reviewer: ```suggestion
You can test your synonyms by adding them directly inline in your token filter definition.
```

[WARNING]
======
This is not recommended for production usage.
A large number of inline synonyms increases cluster size unnecessarily and can lead to performance issues.
======

[discrete]
[[synonyms-synonym-analyzers]]
=== Configure your analyzers to use synonyms sets

[discrete]
[[synonyms-synonym-token-filters]]
==== Configure synonyms token filters

Once your synonyms sets are created, you can start configuring your token filters and analyzers to use them.

{es} uses synonyms as part of the <<analysis-overview,analysis process>>.

> There are two different <<analysis-tokenfilters,token filters>> that can be used for including synonyms:

Reviewer: ```suggestion
You can use two types of <<analysis-tokenfilters,token filters>> to include synonyms:
```

> * <<analysis-synonym-tokenfilter>>: Use it for single word synonyms ("car", "automobile").

Reviewer: ```suggestion
* <<analysis-synonym-tokenfilter>>: For single word synonyms ("car", "automobile").
```

> * <<analysis-synonym-graph-tokenfilter>>: Use it for synonyms that may span multiple words ("hurriedly", "in a hurry").

Reviewer: ```suggestion
* <<analysis-synonym-graph-tokenfilter>>: For synonyms that may span multiple words ("hurriedly", "in a hurry").
```

> 

Reviewer 2: We should phrase this differently, as the way it phrased now looks like synonyms_graph is used only for multiple word:

There are two types of token filters that handle synonyms:
- Synonym graph token filter. It is recommended to always use it, as it can correctly handle multi-word synonyms
- Synonym token filter. Not recommended if you ever need to handle multi-word synonyms.

Author: Thanks for the suggestion! 💯

Synonyms sets created via synonyms APIs can be included into synonym token filters using `synonyms_set` configuration:

> 

Reviewer 2: why "can be included"?
May be we should say like this:

"Use "synonyms_set" option to provide a synonym set created via Synonyms Management APIs"

"Use "synonyms_path" option to provide a synonym file.

"Use "synonyms" option to list inline synonyms.

Author: Sounds more direct, thanks!

[source,JSON]
----
  "filter": {
    "synonyms_filter": {
      "type": "synonym_graph",
      "synonyms_set": "my-synonym-set"
    }
  }
----

The synonym file path needs to be included as part of your token filter using `synonyms_path` configuration:

[source,JSON]
----
  "filter": {
    "synonyms_filter": {
      "type": "synonym_graph",
      "synonyms_path": "analysis/synonym-set.txt"
    }
  }
----

Synonyms can be directly included in your token filter using the `synonyms` configuration:

[source,JSON]
----
  "filter": {
    "synonyms_filter": {
      "type": "synonym_graph",
      "synonyms": ["pc => personal computer", "computer, pc, laptop"]
    }
  }
----

[discrete]
[[synonyms-analizers-configure]]
==== Configure analyzers with synonym token filters

To apply synonyms, you will need to include one of the above synonym token filters into an analyzer, like the following:

[source,JSON]
----
      "analyzer": {

>         "my_analyzer": {

Reviewer 2: should we indicate that this a search analyzer?

Author: I removed the "updateable" references as I think they distracted from the main goal of this section, which was synonyms set configuration. WDYT?

          "type": "custom",
          "tokenizer": "standard",
          "filter": ["lowercase", "synonyms_filter"]
        }
      }
----

Order is important for your token filters.

> {es} will use the token filters preceding the synonym filter to parse the synonym rules in a synonyms set.

Reviewer 2: parse => pass

Author: I copied that directly from the synonyms graph filter, and it didn't sound right. Thanks for catching this!

In the above example, text will be lowercased by the `lowercase` filter before being processed by the `synonyms_filter`.
This means that all the synonyms defined there needs to be in lowercase, or they won't be found by the synonyms filter.

> The synonym rules should not contain words that are removed by a filter that appears after in the chain (like a `stop` filter).

Reviewer: ```suggestion
The synonym rules should not contain words that are removed by a filter that appears later in the chain (like a `stop` filter).
```

Removing a term from a synonym rule means there will be no matching for it at query time.

Because entries in the synonym map cannot have stacked positions, some token filters may cause issues here.
Token filters that produce multiple versions of a token may choose which version of the token to emit when parsing synonyms.
For example, `asciifolding` will only produce the folded version of the token.
Others, like `multiplexer`, `word_delimiter_graph` or `ngram` will throw an error.

If you need to build analyzers that include both multi-token filters and synonym filters, consider using the <<analysis-multiplexer-tokenfilter,multiplexer>> filter, with the multi-token filters in one branch and the synonym filter in the other.

> 

Reviewer 2: I am not sure why we need to duplicate this information here, I think instead of this whole seciton, it is better just to provide a reference to the synonym_filter where all this information is provided in the updated form. Something like this:

Please refer to the synonym_filter documentation to see how this filter interacts with other filters.

Author: I moved this information from the synonyms token filters and included it here. I thought it would be better to consolidate common information in a single place.

[discrete]
[[synonyms-test-analyzer]]
==== Test your analyzer

You can test an analyzer configuration without modifying your index settings.
Use the <<indices-analyze,analyze API>> to test your analyzer chain:

[source,console]
----
GET /_analyze
{
  "tokenizer": "standard",
  "filter" : [
    "lowercase",
    {
      "type": "synonym_graph",
      "synonyms": ["pc => personal computer", "computer, pc, laptop"]
    }
  ],
  "text" : "Check how PC synonyms work"
}
----

[discrete]
[[synonyms-apply-synonyms]]
==== Apply synonyms at index or search time

Analyzers can be applied at <<analysis-index-search-time,index time or search time>>.

You need to decide when to apply your synonyms:

> * Index time: Synonyms are applied when the documents are indexed into Elasticsearch. It is a less flexible alternative, as changing your synonyms imply that your index needs to be <<docs-reindex,reindexed>> to ensure the new synonyms are applied and stored in your index.

Reviewer: ```suggestion
* Index time: Synonyms are applied when the documents are indexed into Elasticsearch. This is a less flexible alternative, as changes to your synonyms require <<docs-reindex,reindexing>>.
```

> * Search time: Synonyms are applied when a search is executed. This is a more flexible approach, as indices don't have to be reindexed, and search analyzers <<indices-reload-analyzers,can be reloaded>> with changes done to synonyms.

Reviewer: ```suggestion
* Search time: Synonyms are applied when a search is executed. This is a more flexible approach, which doesn't require reindexing. With this approach, search analyzers can be <<indices-reload-analyzers,reloaded>> whenever you make changes to your synonyms.
```

> Synonyms sets created using <<synonyms-store-synonyms-api,synonyms API>> can only be used at search time.

Reviewer: ```suggestion
Synonyms sets created using the <<synonyms-store-synonyms-api,synonyms API>> can only be used at search time.
```
